<a href="https://colab.research.google.com/github/jaimeisaac2020/Python-analsisis-basicos/blob/mi-github/pruebas_hipotesis_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [10]:
datos=pd.read_excel('Terapia.xls',sheet_name=1)

In [12]:
datos

n  noche1  noche2  noche3  sexo  edad  terapia1  terapia2  terapia3
0    1     4.0     4.5     5.0     0     1         0         1         0
1    2     5.8     6.4     7.0     1     2         0         1         0
2    3     4.1     4.9     5.1     0     3         0         0         0
3    4     4.3     5.0     5.6     1     1         0         1         1
4    5     5.4     6.1     6.3     0     2         0         0         1
5    6     5.7     6.3     6.8     1     3         1         0         0
6    7     4.3     4.8     5.4     0     1         0         1         1
7    8     7.6     7.9     8.6     1     2         1         0         1
8    9     5.7     6.0     6.7     0     3         1         1         0
9   10     3.8     4.6     4.9     1     1         0         1         1
10  11     7.5     8.0     8.5     0     2         1         0         0
11  12     6.4     7.0     7.3     1     3         1         1         0
12  13     4.0     4.5     5.0     0     1         0         1         1
13  14     7.7     8.1     8.5     1     2         1         1         1
14  15     6.4     6.9     7.3     0     3         1         0         1
15  16     5.9     6.6     7.0     1     1         1         0         1
16  17     7.2     7.8     8.1     0     2         1         1         1
17  18     5.6     6.2     6.8     1     3         1         1         0
18  19     5.7     6.1     6.8     0     1         1         0         0
19  20     7.4     7.6     8.2     1     2         1         1         1

In [13]:
type(datos)


pandas.core.frame.DataFrame

In [15]:
datos.columns


Index(['n', 'noche1', 'noche2', 'noche3', 'sexo', 'edad', 'terapia1',
       'terapia2', 'terapia3'],
      dtype='object')

In [16]:
datos.std()


n           5.916080
noche1      1.317444
noche2      1.235559
noche3      1.252986
sexo        0.512989
edad        0.825578
terapia1    0.502625
terapia2    0.502625
terapia3    0.510418
dtype: float64

#Prueba $Z$ para una media#

La prueba $Z$ se utiliza en el contraste de hipótesis sobre una media cuando la varianza poblacional, $\sigma^2$, es conocida. El estadístico de contraste $Z$ es:
$$
Z=\frac{\bar{X}-\mu_0}{\sigma / \sqrt{n}}
$$
y su distribución es la $\operatorname{Normal}(0,1)$. No existe una función específica de $R$ para calcular el estadístico $Z$, por lo que debemos programarla en $R$.

Por ejemplo, supongamos que disponemos de los siguientes datos referidos a puntuaciones en el "Test PMA de Thurstone" de un grupo de ocho personas con Alzheimer precoz:
$\begin{array}{llllllll}94 & 102 & 88 & 91 & 97 & 94 & 101 & 93\end{array}$

El propósito del análisis es "averiguar si el cociente intelectual medio es menor en la población de la que proceden estas personas que en la población general." Para ello realizamos un contraste unilateral izquierdo con las siguientes hipótesis:
$$
\begin{array}{ll}
H_0: & \mu \geq 100 \\
H_1: & \mu<100
\end{array}
$$  

Como sabemos que el cociente intelectual tiene $\sigma=15$ en la población, podemos hacer el contraste con el estadístico $Z$. Para ello programamos el cálculo de $Z$ del siguiente modo:

In [17]:
import numpy as np

x = np.array([94, 102, 88, 91, 97, 94, 101, 93])
z = (np.mean(x) - 100) / (15 / np.sqrt(8))

print(z)


-0.9428090415820635


El resultado de este cálculo es $Z=-0,94$. Para tomar una decisión sobre $H_0$ podemos emplear dos maneras equivalentes:
- Buscar el punto crítico, $z_\alpha$, que es el valor de $Z$ que deja a su izquierda la probabilidad $\alpha=0,05$. Como vimos en el capítulo anterior, esto puede hacerse con el comando norm.ppf(0.05), donde se ha utilizado un nivel de significación $\alpha=0,05$.  

- Calcular el nivel crítico unilateral izquierdo: $p=P(Z \leq-0,94)$. Para ello empleamos la función pnorm (z), que proporciona la probabilidad acumulada de la distribución normal para el estadístico de contraste definido en $z$.

El punto crítico con $\alpha=0,05$ es $z_\alpha=-1,64$, y el nivel crítico vale $p=0,17$. Con cualquiera de estos métodos llegamos a la decisión de mantener $H_0$. En consecuencia, no podemos concluir que en la población de personas con Alzheimer precoz el cociente intelectual medio sea inferior al de la población general.

In [18]:
from scipy.stats import norm

quantile = norm.ppf(0.05)

print(quantile)



-1.6448536269514729


In [19]:
from scipy.stats import norm

#z =  # Coloca aquí el valor z que desees calcular
cumulative_probability = norm.cdf(z)

print(cumulative_probability)


0.17288929307558015


# Prueba $T$ sobre una media

Cuando el valor de $\sigma^2$ es desconocido, los contrastes sobre medias se realizan con la prueba $T$. En $R$ está implementada la función t.test para realizar el contraste sobre una y dos medias.

Como ejemplo de contraste sobre una media con $\sigma^2$ desconocida, podemos contrastar la hipótesis de que la variable noche1, contenida en datos, tiene media 5 en la población. Las hipótesis son:
$$
\begin{aligned}
& H_0: \mu=5 \\
& H_1: \mu \neq 5
\end{aligned}
$$

El código $R$ para este contraste es:

In [20]:
from scipy import stats



# Realizar la prueba t de dos muestras
t_statistic, p_value = stats.ttest_1samp(datos['noche1'], 5)

# Imprimir el resultado
print("Estadístico t:", t_statistic)
print("Valor p:", p_value)


Estadístico t: 2.4610528516582364
Valor p: 0.02359580334437601


La salida de resultados muestra el estadístico de contraste $(\mathrm{t}=2.4611)$ y el nivel crítico $(\mathrm{p}$-value $=0.0236)$, por lo que la decisión sobre $H_0$ será una u otra dependiendo del nivel de $\alpha$ empleado. La hipótesis nula se mantiene con $\alpha=0,01$ y se rechaza con $\alpha=0,05$.

In [22]:
from scipy import stats



# Prueba t de una muestra contra una media de 5
t_statistic, p_value = stats.ttest_1samp(datos['noche1'], 5)
print("Prueba t de una muestra contra una media de 5:")
print("Estadístico t:", t_statistic)
print("Valor p (bilateral):", p_value)

# Prueba t de una muestra para una media mayor que 5
t_statistic_greater, p_value_greater = stats.ttest_1samp(datos['noche1'], 5, alternative="greater")
print("\nPrueba t de una muestra para una media mayor que 5:")
print("Estadístico t:", t_statistic_greater)
print("Valor p (una cola, mayor que 5):", p_value_greater)

# Prueba t de una muestra para una media menor que 5 con un nivel de confianza del 99%
t_statistic_less, p_value_less = stats.ttest_1samp(datos['noche1'], 5, alternative="less")
alpha = 1 - 0.99
print("\nPrueba t de una muestra para una media menor que 5 con nivel de confianza del 99%:")
print("Estadístico t:", t_statistic_less)
print("Valor p (una cola, menor que 5):", p_value_less)
print("Nivel de confianza del 99% (alpha):", alpha)


Prueba t de una muestra contra una media de 5:
Estadístico t: 2.4610528516582364
Valor p (bilateral): 0.02359580334437601

Prueba t de una muestra para una media mayor que 5:
Estadístico t: 2.4610528516582364
Valor p (una cola, mayor que 5): 0.011797901672188004

Prueba t de una muestra para una media menor que 5 con nivel de confianza del 99%:
Estadístico t: 2.4610528516582364
Valor p (una cola, menor que 5): 0.988202098327812
Nivel de confianza del 99% (alpha): 0.010000000000000009


#Contraste sobre dos medias y dos varianzas independientes

En un diseño inter-sujetos, disponemos de dos grupos formados por diferentes personas y queremos comparar las medias de los grupos. A este análisis se le denomina contraste de dos medias independientes. El contraste tiene dos versiones: una en la que se asumen varianzas iguales en los dos grupos comparados y otra en la que las varianzas de los grupos son distintas. En ambos casos se calcula un estadístico $T$, pero el modo de calcularlo y sus grados de libertad son distintos.

Para escoger una de estas versiones de la prueba $T$ (varianzas poblacionales iguales o distintas), debemos empezar realizando un Contraste sobre la igualdad de las varianzas de los grupos, cuyas hipótesis son:
$$
\begin{aligned}
& H_0: \sigma_1^2=\sigma_2^2 \\
& H_1: \sigma_1^2 \neq \sigma_2^2
\end{aligned}
$$

En función del resultado del contraste de varianzas, nos decantaremos por la prueba $T$ para dos medias independientes con varianzas iguales o con varianzas distintas.

Supongamos que queremos comparar las medias de noche1 en varones y mujeres. El primer paso es utilizar la función var.test para realizar el contraste de igualdad de varianzas. El argumento que le pasamos a var.test es la fórmula noche1 sexo, que significa que la variable dependiente es noche1 y, por tanto, vamos a comparar la varianza de noche1 en los grupos formados por la variable independiente sexo.